File: cp-library-rs/src/segment_tree.rs
//! セグメント木

use crate::monoid::Monoid;
use std::fmt::{self, Debug};
use std::ops::{
    Bound::{Excluded, Included, Unbounded},
    Deref, DerefMut, Index, RangeBounds,
};

/// # SegmentTree (Monoid)
/// - 抽象化セグメント木
pub struct SegmentTree<M: Monoid> {
    pub size: usize,
    offset: usize,
    data: Vec<M::Val>,
}

impl<M: Monoid> Index<usize> for SegmentTree<M> {
    type Output = M::Val;
    fn index(&self, idx: usize) -> &Self::Output {
        &self.data[self.offset + idx]
    }
}

impl<M: Monoid> SegmentTree<M> {
    #[inline]
    fn parse_range<R: RangeBounds<usize>>(&self, range: &R) -> Option<(usize, usize)> {
        let start = match range.start_bound() {
            Unbounded => 0,
            Excluded(&v) => v + 1,
            Included(&v) => v,
        };
        let end = match range.end_bound() {
            Unbounded => self.size,
            Excluded(&v) => v,
            Included(&v) => v + 1,
        };
        if start <= end && end <= self.size {
            Some((start, end))
        } else {
            None
        }
    }

    /// セグメント木を初期化する
    pub fn new(N: usize) -> Self {
        let offset = N;

        Self {
            size: N,
            offset,
            data: vec![M::id(); offset << 1],
        }
    }

    pub fn update(&mut self, index: usize, value: M::Val) {
        let mut i = index + self.offset;
        self.data[i] = value;
        while i > 1 {
            i >>= 1;
            let lch = i << 1;
            self.data[i] = M::op(&self.data[lch], &self.data[lch + 1]);
        }
    }

    /// 可変な参照を返す
    pub fn get_mut(&mut self, i: usize) -> Option<ValMut<'_, M>> {
        if i < self.offset {
            let default = self.index(i).clone();
            Some(ValMut {
                segtree: self,
                idx: i,
                new_val: default,
            })
        } else {
            None
        }
    }

    /// 区間`range`の集約を行う
    pub fn get_range<R: RangeBounds<usize> + Debug>(&self, range: R) -> M::Val {
        let (start, end) = match self.parse_range(&range) {
            Some(r) => r,
            None => panic!("The given range is wrong: {:?}", range),
        };
        // 値の取得
        let mut l = self.offset + start;
        let mut r = self.offset + end;
        let (mut res_l, mut res_r) = (M::id(), M::id());

        while l < r {
            if l & 1 == 1 {
                res_l = M::op(&res_l, &self.data[l]);
                l += 1;
            }
            if r & 1 == 1 {
                r -= 1;
                res_r = M::op(&self.data[r], &res_r);
            }
            l >>= 1;
            r >>= 1;
        }

        M::op(&res_l, &res_r)
    }
}

impl<M: Monoid> From<&Vec<M::Val>> for SegmentTree<M> {
    fn from(src: &Vec<M::Val>) -> Self {
        let mut seg = Self::new(src.len());
        for (i, v) in src.iter().enumerate() {
            seg.data[seg.offset + i] = v.clone();
        }
        for i in (0..seg.offset).rev() {
            let lch = i << 1;
            seg.data[i] = M::op(&seg.data[lch], &seg.data[lch + 1]);
        }
        seg
    }
}

impl<M: Monoid> Debug for SegmentTree<M> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "SegmentTree {{ [").ok();
        for i in 0..self.size {
            if i + 1 < self.size {
                write!(f, "{:?}, ", self.data[self.offset + i]).ok();
            } else {
                write!(f, "{:?}", self.data[self.offset + i]).ok();
            }
        }
        write!(f, "] }}")
    }
}

pub struct ValMut<'a, M: 'a + Monoid> {
    segtree: &'a mut SegmentTree<M>,
    idx: usize,
    new_val: M::Val,
}

impl<M: Monoid> Debug for ValMut<'_, M> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_tuple("ValMut").field(&self.new_val).finish()
    }
}

impl<M: Monoid> Drop for ValMut<'_, M> {
    fn drop(&mut self) {
        self.segtree.update(self.idx, self.new_val.clone());
    }
}

impl<M: Monoid> Deref for ValMut<'_, M> {
    type Target = M::Val;
    fn deref(&self) -> &Self::Target {
        &self.new_val
    }
}

impl<M: Monoid> DerefMut for ValMut<'_, M> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.new_val
    }
}

impl<M> SegmentTree<M>
where
    M: Monoid,
    M::Val: Debug,
{
    /// セグ木を簡易的に表示する
    /// **サイズが2べきのときのみ**
    pub fn show(&self) {
        #[cfg(debug_assertions)]
        {
            let mut i = 1;
            let mut w = 1;
            while i + w <= 2 * self.offset {
                eprintln!("{:?}", &self.data[i..i + w]);
                i += w;
                w <<= 1;
            }
            eprintln!();
        }
    }
}

File: cp-library-py/cp_library_py/segment_tree.py
from typing import Callable, Generic, TypeVar

T = TypeVar("T")


class SegmentTree(Generic[T]):
    """セグメント木"""

    def __init__(self, n: int, id_: Callable[[], T], op: Callable[[T, T], T]) -> None:
        """セグメント木を初期化する

        Args:
            n (int): 要素数
            id_ (Callable[[], T]): 単位元を返す関数
            _op (Callable[[T, T], T]): 二項演算を行う関数
        """
        self._size = n
        self._offset = 1 << (n - 1).bit_length()
        self._id = id_
        self._op = op
        self._data = [id_() for _ in range(self._offset * 2)]

    @classmethod
    def from_array(cls, array: list[T], id_: Callable[[], T], op: Callable[[T, T], T]) -> "SegmentTree[T]":
        """配列からセグメント木を構築する

        Args:
            array (list[T]): 配列
            id_ (Callable[[], T]): 単位元を返す関数
            _op (Callable[[T, T], T]): 二項演算を行う関数

        Returns:
            SegmentTree[T]: 構築したセグメント木
        """
        seg = cls(len(array), id_, op)
        seg._data[seg._offset: seg._offset + len(array)] = array

        for i in range(seg._offset - 1, 0, -1):
            seg._data[i] = seg._op(seg._data[i * 2], seg._data[i * 2 + 1])

        return seg

    def update(self, i: int, x: T) -> None:
        """i 番目の要素を x に更新する

        Args:
            i (int): 更新する要素のインデックス
            x (T): 更新する値
        """
        i += self._offset
        self._data[i] = x
        while i > 1:
            i >>= 1
            self._data[i] = self._op(self._data[i * 2], self._data[i * 2 + 1])

    def get_range(self, l: int, r: int) -> T:
        """区間 [l,r) の演算結果を返す

        Args:
            l (int): 区間の左端
            r (int): 区間の右端

        Returns:
            T: 演算結果
        """
        l += self._offset
        r += self._offset
        res_l = self._id()
        res_r = self._id()

        while l < r:
            if l & 1:
                res_l = self._op(res_l, self._data[l])
                l += 1
            if r & 1:
                r -= 1
                res_r = self._op(self._data[r], res_r)
            l >>= 1
            r >>= 1

        return self._op(res_l, res_r)

    def __getitem__(self, i: int) -> T:
        """i 番目の要素を取得する

        Args:
            i (int): 取得する要素のインデックス

        Returns:
            T: 取得した要素
        """
        return self._data[i + self._offset]

    def __setitem__(self, i: int, x: T) -> None:
        """i 番目の要素を x に更新する

        Args:
            i (int): 更新する要素のインデックス
            x (T): 更新する値
        """
        self.update(i, x)
